### PR TITLE
mobile: Use ktfmt to format Kotlin code

### PR DESCRIPTION
**Pull request by fredyw**
This PR creates a wrapper script to format Kotlin code with `ktfmt` instead of `ktlint` since `ktfmt` is more widely used and the existing build rule with `ktlint` to format Kotlin code does not work. This PR also formats Kotlin code by running `./tools/check_format fix`.

Risk Level: n/a
Testing: manual & CI
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

**Review by abeyad — DISMISSED** [review]
awesome! thanks for doing this!

**Review by abeyad — DISMISSED** [review]

**Review by abeyad — DISMISSED** [review]

File: mobile/tools/ktfmt.sh
Comment (phlax) [inline mobile/tools/ktfmt.sh:3]: @abeyad @fredyw before we land this can we figure out best way forward re installation

iiuc this will install the jdk every time it runs

Comment (abeyad) [reply mobile/tools/ktfmt.sh:3]: we can check if it's already installed by something like `javac -version || command_to_install` ?

Comment (phlax) [reply mobile/tools/ktfmt.sh:3]: it was more the question of whether its already installed and whether we would be better placing any installation in the build image

my overall pref would be for a bazel target and no host install - but that is probably a bit more work

Comment (fredyw) [reply mobile/tools/ktfmt.sh:3]: The script will actually not download the JDK again if it already exists.

```
if [[ ! -f "${java}" ]]; then
  download_jdk
fi
```

The nice thing with this approach is when running on non-CI, the developers will not need to worry about installing the JDK since it will automatically install it and this step will only happen once.

I can also update the CI to install the JDK in the Mobile image and update this script to use the JDK from the CI if it exists. How does that sound?

Comment (phlax) [reply mobile/tools/ktfmt.sh:3]: as a rule we try to avoid host bins - and if we do need them the place to put them generally is in the build image

i think this encumbers devs because they are already (potentially) download 2 different versions of the jdk so this just adds another

that said - i dont want to block so ill leave it to @abeyad to figure out best way forward

Comment (phlax) [reply mobile/tools/ktfmt.sh:3]: one other thought on this - if the jdk is only required to compile the linter then the definitely preferable option (without using bazel) would be to do that in a docker layer and then its not going to make the image too much larger etc and altho not hermetic etc - at least the version/build is determinate

Comment (abeyad) [reply mobile/tools/ktfmt.sh:3]: my understanding is that what Fredy is doing here doesn't increase the number of jdks installed any more than shoving the same functionality into a bazel rule, so from that point of view, i'm not sure if there's an added benefit in putting the "pull jdk version to run ktfmt" logic into a bazel rule over the script.  note that this jdk is not used to actually compile the mobile build's files (the bazel jdk is used for that).

> the definitely preferable option (without using bazel) would be to do that in a docker layer 
not sure what you mean here, what should be done in the docker layer?

Comment (phlax) [reply mobile/tools/ktfmt.sh:3]: >  doesn't increase the number of jdks installed any more than shoving the same functionality into a bazel rule, 

how so? bazel already pulls in a jdk iiuc in addition to the one that is installed in the mobile build image

we want to be reducing the number of these (and the number of versions used) not increase them

Comment (phlax) [reply mobile/tools/ktfmt.sh:3]: also - with these kind of binaries - my plan (which is partially implemented) is to publish the static binaries ourselves 

in the case of clang-tidy/format this reduces the download requirements for jobs that only require these from ~4GB to <100MB

Comment (fredyw) [reply mobile/tools/ktfmt.sh:3]: > if the jdk is only required to compile the linter

The JDK is actually only used to **run** the formatter and *not* to compile it. The JAR has already been compiled when we download it.

> at least the version/build is determinate

The script will download a specific version of the JAR as well as the JDK and we'll also verify the checksums, so it will also be deterministic.

My only hesitation from putting this into the Docker image is there will be two JDKs, one that will be used for the compilation (JDK 8) for Android and the other one that's only used as a runner (JDK 11) for `ktfmt` and we will need to do `update-java-alternatives` to make sure we don't default it into the wrong version. There's nothing wrong with this approach, but it simply complicates the process a bit more. When doing the development (not on CI), we also want to make this step as seamless as possible and not having to require the users to deal with choosing the right JDK version and installing it.

Comment (phlax) [reply mobile/tools/ktfmt.sh:3]: > there will be two JDKs, ...

3 if you include what bazel has setup (not sure if its pulled specifically in this ci)

this is the point im making - it would be much better to consolidate this and not have users download so much - and complicating the required dependencies.

> ... one that will be used for the compilation (JDK 8) for Android and the other one that's only used as a runner (JDK 11) for ktfmt and we will need to do update-java-alternatives to make sure we don't default it into the wrong version. There's nothing wrong with this approach, but it simply complicates the process a bit more. 

switching version seems simpler to me than downloading a complete new toolchain

> When doing the development (not on CI), we also want to make this step as seamless as possible and not having to require the users to deal with choosing the right JDK version and installing it.

agreed - hence my pushback

iirc from building the build images - the apt repo is not rock solid either, so my guess is that will also apply to downloading the tarball

re it wont download if its there already - it will in ci im guessing

as i said i wont push back any further, i think it will affect a very small subset of devs, and we can deal with ci issues if they come up. But i do want to make the point that it is the opposite of what we are trying to do in terms of deps, and gently suggest that it may not really be necessary, at least not like this

Comment (fredyw) [reply mobile/tools/ktfmt.sh:3]: SGTM. I added a TODO and will update the CI's Docker Image to a follow up PR. Thanks for the suggestions!

Comment (abeyad) [reply mobile/tools/ktfmt.sh:3]: a follow up on updating the docker image sounds good, thanks @fredyw and @phlax !

**Review by abeyad — APPROVED** [review]

**Pull request merged**
